Core/GDCore/Events/CodeGeneration/EffectsCodeGenerator.cpp

```diff
@@ -15,16 +15,45 @@
 #include "GDCore/Project/Layout.h"
 #include "GDCore/Project/Object.h"
 #include "GDCore/Project/Project.h"
+#include "GDCore/IDE/ProjectBrowserHelper.h"
 
 namespace gd {
 
-void ExposeProjectEffects(
-    const gd::Project& project,
-    const std::function<void(const gd::Effect& effect)>& worker) {
+void EffectsCodeGenerator::DoVisitObject(gd::Object &object) {
+  auto &effects = object.GetEffects();
+  for (std::size_t e = 0; e < effects.GetEffectsCount(); e++) {
+    auto &effect = effects.GetEffect(e);
+    AddEffectIncludeFiles(effect);
+  }
+};
+
+void EffectsCodeGenerator::AddEffectIncludeFiles(const gd::Effect &effect) {
+  // TODO: this browse all the extensions every time we're trying to find
+  // a new effect. Might be a good idea to rework MetadataProvider to be
+  // faster (not sure if it is a bottleneck at all though - but could be

```

Comment:
Yes, so that access is O(1) via a map rather than iterating on extensions (but we have to be sure it's always up to date with extensions, including when some are added/removed/changed). Can be done later.

Reply:
Oh this is an existing comment. Still useful!